shell/app/electron_crash_reporter_client.cc

```diff
@@ -0,0 +1,182 @@
+// Copyright 2013 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "shell/app/electron_crash_reporter_client.h"
+
+#include "base/command_line.h"
+#include "base/environment.h"
+#include "base/files/file_path.h"
+#include "base/logging.h"
+#include "base/path_service.h"
+#include "base/strings/string_split.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/threading/thread_restrictions.h"
+#include "build/build_config.h"
+#include "chrome/common/chrome_paths.h"
+#include "components/crash/core/common/crash_keys.h"
+#include "components/upload_list/crash_upload_list.h"
+#include "content/public/common/content_switches.h"
+#include "electron/electron_version.h"
+#include "services/service_manager/embedder/switches.h"
+
+#if defined(OS_POSIX) && !defined(OS_MACOSX)
+#include "components/version_info/version_info_values.h"
+#endif
+
+#if defined(OS_POSIX)
+#include "base/debug/dump_without_crashing.h"
+#endif
+
+namespace {
+
+ElectronCrashReporterClient* Instance() {
+  static base::NoDestructor<ElectronCrashReporterClient> crash_client;
+  return crash_client.get();
+}
+
+}  // namespace
+
+// static
+void ElectronCrashReporterClient::Create() {
+  crash_reporter::SetCrashReporterClient(Instance());
+
+  // By setting the BREAKPAD_DUMP_LOCATION environment variable, an alternate
+  // location to write crash dumps can be set.
+  std::unique_ptr<base::Environment> env(base::Environment::Create());
+  std::string alternate_crash_dump_location;
+  base::FilePath crash_dumps_dir_path;
+  if (env->GetVar("BREAKPAD_DUMP_LOCATION", &alternate_crash_dump_location)) {
```

Comment:
This is copied from Chrome, and it seemed reasonable enough to support. The main reason this is relevant in Chrome is that crashpad initialization happens very early (before commandline parsing even) so an environment variable is sort of the only reasonable way to customize this. We give that up in Electron and initialize a ton of stuff before the crash reporter since the crash reporter is initialized by JS, so this is i suppose less relevant for us.

I think it's useful beyond `app.setPath('crashDumps')` because having it in core would mean it's the same for every app. It is a bit of a misnomer though since it's not actually breakpad-specific.

I've also been pondering having a way to initialize the crash reporter extra-early even in the main process by statically encoding the necessary information for initialization (chiefly submit url and product name/version), e.g. in the package.json file. If that were enabled then `app.setPath` wouldn't be able to be called early enough.

WDYT? Perhaps we should drop this code for now? Or maybe it's worth keeping?

Reply:
Thanks for the context! I think its okay to have this shortcut, might help with some cases of debugging (ex: app has published itself with hardcoded crash directory but wants to test with some other path by asking user to set this env variable).

> I've also been pondering having a way to initialize the crash reporter extra-early even in the main process 

What would be the advantage if we do it ?

Reply:
The advantage would be that we would be able to catch crashes that happened before the app's JS was called. There's a lot of setup that happens prior to JS being invoked, and currently if any of it crashes then `crashReporter` won't be able to know about it. I'm not sure if any child processes are booted prior to JS being invoked (I don't think they are...?) but if they were, then they wouldn't have crash reporter coverage either.